packages/labs/analyzer/src/lib/program-context.ts

```diff
@@ -0,0 +1,478 @@
+    const jsdocType = ts.getJSDocType(node);
+    return this.convertType(
+      jsdocType
+        ? this.checker.getTypeFromTypeNode(jsdocType)
+        : this.checker.getTypeAtLocation(node),
+      node
+    );
+  }
+
+  /**
+   * Returns the module specifier for a declaration if it was imported,
+   * or `undefined` if the declaration was not imported.
+   */
+  getImportModuleSpecifier(declaration: ts.Node): string | undefined {
+    // TODO(kschaaf) support the various import syntaxes, e.g. `import {foo as bar} from 'baz'`
+    if (
+      ts.isImportSpecifier(declaration) &&
+      ts.isNamedImports(declaration.parent) &&
+      ts.isImportClause(declaration.parent.parent) &&
+      ts.isImportDeclaration(declaration.parent.parent.parent)
+    ) {
+      const module = declaration.parent.parent.parent.moduleSpecifier
+        .getText()
+        // Remove quotes
+        .slice(1, -1);
+      return module;
+    }
+    return undefined;
+  }
+
+  /**
+   * Returns an analyzer `Reference` object for a named identifier at the given
+   * AST location.
+   *
+   * If the symbol's declaration was imported, the Reference will be based on
+   * the import's module specifier; otherwise the Reference will point to the
+   * current module being analyzed.
+   */
+  getReferenceForSymbolName(name: string, location: ts.Node): Reference {
+    const symbol = this.getSymbolForName(name, location);
+    if (this.currentModule === undefined) {
+      throw new Error(`Internal error: expected currentModule to be set`);
+    }
+    const declaration = symbol?.declarations?.[0];
+    if (declaration === undefined) {
+      throw new DiagnosticsError(
+        location,
+        `Could not find declaration for symbol '${name}'`
+      );
+    }
+    if (declaration.getSourceFile() !== location.getSourceFile()) {
+      // If the reference declaration doesn't exist in this module, it must have
+      // been a global (whose declaration is in an ambient .d.ts file)
+      return new Reference({
+        name,
+        isGlobal: true,
+      });
+    } else {
+      const module = this.getImportModuleSpecifier(declaration);
+      if (module) {
+        // The symbol was imported; check whether it is package local or external
+        if (module[0] === '.') {
+          // Relative import from this package: use the current package and
+          // module path relative to this module
+          return new Reference({
+            name,
+            package: this.package,
+            module: path.join(path.dirname(this.currentModule.jsPath), module),
+          });
+        } else {
+          // External import: extract the npm package (taking care to respect
+          // npm orgs) and module specifier (if any)
+          const info = module.match(
+            /^(?<package>(@\w+\/\w+)|\w+)\/?(?<module>.*)$/
+          );
+          if (!info || !info.groups) {
+            throw new DiagnosticsError(
```

Comment:
It think the absolute path case (`import '/...'`) will fall through to here and it might be a confusing error message.

Reply:
Added support for absolute and href specifiers